### PR TITLE
Development/delete and update options for schedule

### LINKED ISSUE #325: Rework schedule service
Complete rework for Schedule Service

### PR DESCRIPTION
**Pull request by CptHouston**
-- update an event
-- update events range
-- delete event occurrence and related events 

relevant updates are in
- SchedulesController.cs
- ScheduleService.cs

### CODE REVIEW

**Comment by CptHouston** [comment]
should be merged after https://github.com/ita-social-projects/WhatBackend/pull/358

File: CharlieBackend.Api/Controllers/SchedulesController.cs [hunk 1]
```diff
@@ -37,74 +38,125 @@ public SchedulesController(IScheduleService scheduleService)
...
 
             return foundSchedules.ToActionResult();
         }
 
         /// <summary>
-        /// Updates shedule
+        /// Updates a single event
         /// </summary>
+        /// <remarks>
+        /// Old instance os event occurance is replaced with the new one (id is the same).
```
Comment (filinskyi) [inline CharlieBackend.Api/Controllers/SchedulesController.cs]: > Old instance **os** of event **occurance** is replaced with the new one (id is the same).

replace bold words:
os -> of 
occurance -> occurrence

File: CharlieBackend.Api/Controllers/SchedulesController.cs [hunk 2]
```diff
@@ -37,74 +38,125 @@ public SchedulesController(IScheduleService scheduleService)
...
+            var foundSchedules = await _scheduleService.UpdateEventOccurrenceById(eventOccurrenceID, updateOccurrenceRequest);
 
             return foundSchedules.ToActionResult();
         }
 
         /// <summary>
         /// Deletes exact schedule
         /// </summary>
+        /// <remarks>
+        /// Removes cheduled events related to specified EventOccurrence, updates start and finish date accordingle
```
Comment (filinskyi) [inline CharlieBackend.Api/Controllers/SchedulesController.cs]: > Removes **cheduled** events related to specified EventOccurrence, updates start and finish date **accordingle**

replace bold words:
cheduled -> scheduled
accordingle -> accordingly

File: CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            DateTime startDate = new DateTime(_source.EventStart.Year, _source.EventStart.Month, 7 * (int)_pattern.Index,
+                _source.EventStart.Hour, _source.EventStart.Minute, _source.EventStart.Second);
+
+            int offset = _pattern.DaysOfWeek[index] == DayOfWeek.Sunday ? 7 : (int)_pattern.DaysOfWeek[index];
+            int startDay = (int)startDate.DayOfWeek;
+
+            return startDate.AddDays(startDay <= offset
+                       ? -1 * (startDay + 7 - offset)
+                       : -1 * (startDay - offset));
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs]: magic number 7, may be change it to some constant or static field ,
as you doing here :
image

File: CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+            startDate = new DateTime(startDate.Year, startDate.Month, 7 * (int)_pattern.Index,
+                startDate.Hour, startDate.Minute, startDate.Second);
+
+            int offset = _pattern.DaysOfWeek[index] == DayOfWeek.Sunday ? 7 : (int)_pattern.DaysOfWeek[index];
+            int startDay = (int)startDate.DayOfWeek;
+
+            startDate = startDate.AddDays(startDay <= offset
+                       ? -1 * (startDay + 7 - offset)
+                       : -1 * (startDay - offset));
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs]: same suggestion as in upper comment

File: CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerWeekly.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            _index = 7;
+        }
+
+        protected override DateTime GetStartDate(int index)
+        {
+            DateTime result = _source.EventStart;
+            
+            return result.AddDays(result.DayOfWeek <= _pattern.DaysOfWeek[index]
+                    ? _pattern.DaysOfWeek[index] - result.DayOfWeek
+                    : 7 - (int)result.DayOfWeek + (int)_pattern.DaysOfWeek[index]);
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerWeekly.cs:23]: same suggestion as in upper comment

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+                        error = "Target days not provided";
+                    }
+                    break;
+                default:
+                    error = "Pattern type not supported";
+                    break;
+            }
+
+            return error;
+        }
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: these checks will override previous messages.
as example if all of them are valid "error" message will contain only the last one message.

or return in each "if" "case"
or if you want to do all checks and accumulate all messages than use StringBuilder.

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+            {
+                error = $"No such theme id={request.ThemeId.Value}";
+            }
+
+            if (request.EventEnd.HasValue && request.EventStart.HasValue && (request.EventEnd < request.EventStart))
+            {
+                error = $"StartDate must be less then FinisDate";
+            }
+
+            return error;
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: same case as previous comment

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+                error = "Theme does not exist";
+            }
+
+            if (request.StartDate.HasValue && request.FinishDate.HasValue && (request.StartDate < request.FinishDate))
+            {
+                error = $"StartDate must be less then FinisDate";
+            }
+
+            return error;
+        }
```
Comment (filinskyi) [inline CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: same case as previous comment

File: CharlieBackend.Core/Entities/EventOccurrence.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.ComponentModel.DataAnnotations;   
+
+namespace CharlieBackend.Core.Entities
+{
+    public partial class EventOccurrence : BaseEntity
```
Comment (Hlib-mats) [inline CharlieBackend.Core/Entities/EventOccurrence.cs]: why does this class partial?

File: CharlieBackend.Core/Entities/Theme.cs [hunk 10]
```diff
@@ -12,5 +12,7 @@ public Theme()
         public string Name { get; set; }
 
         public virtual ICollection<Lesson> Lessons { get; set; }
+
+        public virtual ICollection<ScheduledEvent> ScheduledEvents { get; set; }
```
Comment (Hlib-mats) [inline CharlieBackend.Core/Entities/Theme.cs:16]: Does theme actually need schedule?

**Review by Hlib-mats — APPROVED** [review]

File: CharlieBackend.Api/Controllers/SchedulesController.cs (hunk 1 above)
Comment (CptHouston) [reply CharlieBackend.Api/Controllers/SchedulesController.cs]: updated

File: CharlieBackend.Api/Controllers/SchedulesController.cs (hunk 2 above)
Comment (CptHouston) [reply CharlieBackend.Api/Controllers/SchedulesController.cs]: updated

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs (hunk 6 above)
Comment (CptHouston) [reply CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: accumulating errors is a good idea
updated

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs (hunk 7 above)
Comment (CptHouston) [reply CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: updated

File: CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs (hunk 8 above)
Comment (CptHouston) [reply CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs]: updated

File: CharlieBackend.Core/Entities/EventOccurrence.cs (hunk 9 above)
Comment (CptHouston) [reply CharlieBackend.Core/Entities/EventOccurrence.cs]: oversight from previous built
updated

File: CharlieBackend.Core/Entities/Theme.cs (hunk 10 above)
Comment (CptHouston) [reply CharlieBackend.Core/Entities/Theme.cs:16]: field added as a part of the foreign key constraint

File: CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs (hunk 4 above)
Comment (CptHouston) [reply CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs]: updated

File: CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerWeekly.cs (hunk 5 above)
Comment (CptHouston) [reply CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerWeekly.cs:23]: updated

**Comment by CptHouston** [comment]
pr transferred https://github.com/ita-social-projects/WhatBackend/pull/402

**Pull request closed without merging**